src/Microsoft.AspNetCore.Rewrite/UrlRewrite/InputParser.cs

```diff
@@ -0,0 +1,212 @@
+                            {
+                                var pattern = ParseString(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    var str = pattern.Evaluate(ctx, ruleMatch, condMatch);
+                                    return str.ToLowerInvariant();
+                                }));
+
+                                // at this point, we expect our context to be on the ending closing brace,
+                                // because the ParseString() call will increment the context until it 
+                                // has processed the new string.
+                                if (context.Current != CloseBrace)
+                                {
+                                    throw new FormatException();
+                                }
+                            }
+                            break;
+                        case "UrlDecode":
+                            {
+                                throw new NotImplementedException("Do no support UrlDecoding.");
+                            }
+                        case "UrlEncode":
+                            {
+                                var pattern = ParseString(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    var str = pattern.Evaluate(ctx, ruleMatch, condMatch);
+                                    return UrlEncoder.Default.Encode(str);
+                                }));
+
+                                if (context.Current != CloseBrace)
+                                {
+                                    throw new FormatException();
+                                }
+                            }
+                            break;
+                        case "R":
+                            {
+                                var index = GetBackReferenceIndex(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    return ruleMatch?.Groups[index]?.Value;
+                                }));
+                                return;
+                            }
+                        case "C":
+                            {
+                                var index = GetBackReferenceIndex(context);
+                                results.Add(new PatternSegment((ctx, ruleMatch, condMatch) =>
+                                {
+                                    return condMatch?.Groups[index]?.Value;
+                                }));
+                                return;
+                            }
+                        default:
+                            throw new FormatException("Unrecognized parameter.");
```

Comment:
What parameter is unrecognized? All exceptions should have a context and they should be actionable by someone who doesn't know the implemention